### PR TITLE
[posix] stop setting up and de-initializing uninitialized modules when dry run

### PR DESCRIPTION
**Pull request by jinran-google**
Initialization of these modules is not called when dry run, so setup and de-initialization should not be called either.

### CODE REVIEW

File: src/posix/platform/system.cpp [hunk 1]
```diff
@@ -196,7 +196,10 @@ otInstance *otSysInit(otPlatformConfig *aPlatformConfig)
     gInstance = otInstanceInitSingle();
     OT_ASSERT(gInstance != nullptr);
 
-    platformSetUp();
+    if (!gDryRun)
+    {
+        platformSetUp();
```
Comment (zhanglongxia) [inline src/posix/platform/system.cpp:201]: Why do we need to disable `platformSetUp()` when the dry run is enabled? Do we also need to disable `platformTearDown()` when the dry run is enabled?

Comment (jinran-google) [reply src/posix/platform/system.cpp:201]: Some uninitialized data might be used when setting up these modules. E.g. the interface name when creating the daemon `.lock` file.

I changed the PR to disable `platformTearDown()` as well.

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/posix/platform/system.cpp (hunk 1 above)
Comment (simonlingoogle) [inline src/posix/platform/system.cpp:201]: move the logic to `platformSetUp` ?

File: src/posix/platform/system.cpp [hunk 2]
```diff
@@ -247,13 +254,19 @@ void platformDeinit(void)
...
 
 void otSysDeinit(void)
 {
     OT_ASSERT(gInstance != nullptr);
 
-    platformTearDown();
+    if (!gDryRun)
+    {
+        platformTearDown();
+    }
```
Comment (simonlingoogle) [inline src/posix/platform/system.cpp:269]: move the logic to platformTearDown ?

File: src/posix/platform/settings.cpp [hunk 3]
```diff
@@ -225,7 +225,10 @@ void otPlatSettingsDeinit(otInstance *aInstance)
     OT_UNUSED_VARIABLE(aInstance);
 
 #if OPENTHREAD_POSIX_CONFIG_SECURE_SETTINGS_ENABLE
-    otPosixSecureSettingsDeinit(aInstance);
+    if (!IsSystemDryRun())
```
[the remark this answers is not in the document]
Comment (jinran-google) [reply src/posix/platform/settings.cpp]: Sounds more reasonable. Done.

File: src/posix/platform/system.cpp (hunk 1 above)
Comment (jinran-google) [reply src/posix/platform/system.cpp]: Done.

File: src/posix/platform/system.cpp (hunk 2 above)
Comment (jinran-google) [reply src/posix/platform/system.cpp]: Done.

**Review by simonlingoogle — APPROVED** [review]
LGTM 👍 

Thanks

**Review by bukepo — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**